# Aula 4 - Manipulação de df: groupby e merge


### Objetivos

Apresentar como unir dataframes e realizar cálculos com dados agrupados

____________________________

### Habilidades a serem desenvolvidas nessa aula

Ao final da aula o aluno deve:

- Saber como concatenar dataframes,
- Conseguir agrupar os dados e aplicar vários métodos à eles


____
____
____

## Titanic

O arquivo que usaremos hoje é relativo ao Titanic! Essa é uma das bases mais famosas de ciência de dados. Você pode saber mais sobre estes dados [clicando aqui!](https://www.kaggle.com/c/titanic)

In [15]:
import pandas as pd

In [16]:
# lê dataframe do arquivo titanic.csv 
df = pd.read_csv("data/titanic.csv")

E se quisessemos calcular a média de Fare por Pclasse utilizando apenas o que aprendemos até agora?

In [17]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [24]:
df[df['Pclass']==1].Fare#.mean()

1      71.2833
3      53.1000
6      51.8625
11     26.5500
23     35.5000
        ...   
871    52.5542
872     5.0000
879    83.1583
887    30.0000
889    30.0000
Name: Fare, Length: 216, dtype: float64

In [20]:
df[df['Pclass']==2].Fare.mean()

20.662183152173913

In [21]:
df[df['Pclass']==3].Fare.mean()

13.675550101832993

Ou de forma mais automática:

In [22]:
df.Pclass.sort_values().unique()

array([1, 2, 3])

In [23]:
for i in df.Pclass.sort_values().unique():
    print(f'Ticket médio da classe {i}: ', df[df['Pclass']==i].Fare.mean())

Ticket médio da classe 1:  84.1546875
Ticket médio da classe 2:  20.662183152173913
Ticket médio da classe 3:  13.675550101832993


E se quisessemos calcular a média por Pclass e Sex?

### Groupby
Assim como no SQL, no pandas também temos um método com o qual podemos agregar os dados. O `groupby` primeiro separa nossos dados em grupos definidos dentro do método,  aplica um tipo de operação usando agregação, transformação, filtragem ou até uma função própria e, por fim, junta os resultados encontrados.
<br>

<img src="groupby.png"  style="width: 700px" >

Exemplo de aplicação da função de agregação `mean`
<br><br><br>

Utilizar o `groupby` é o mesmo que fazer a sequência:

   1. Dividir os dados em grupos utilizando um critério
    
   2. Aplicar uma função em cada um dos grupos separadamente
    
   3. Combinar o resultado em uma estrutura de dados

#### Funções de agregação
Com essas funções podemos aplicar operações estatísticas nos nossos dados. Exemplos:<br>
`mean`, `std`, `max`, `min`, `count`, `sum`, `var`. <br>
Quando queremos aplicar apenas uma dessas operações podemos chamá-las diretamente após o `groupby`:


In [36]:
df.groupby(['Pclass']).Fare.mean()

Pclass
1   84.15
2   20.66
3   13.68
Name: Fare, dtype: float64

In [35]:
df.groupby(['Pclass'])['Fare'].mean()

Pclass
1   84.15
2   20.66
3   13.68
Name: Fare, dtype: float64

In [34]:
df.groupby(['Pclass'])[['Fare']].mean()

Fare
Pclass      
1      84.15
2      20.66
3      13.68

In [32]:
pd.options.display.float_format= "{:,.2f}".format

In [33]:
# Agrupa por Pclass e Sex e calcula a média de cada grupo
df.groupby(['Pclass', 'Sex']).mean()


PassengerId  Survived   Age  SibSp  Parch   Fare
Pclass Sex                                                     
1      female       469.21      0.97 34.61   0.55   0.46 106.13
       male         455.73      0.37 41.28   0.31   0.28  67.23
2      female       443.11      0.92 28.72   0.49   0.61  21.97
       male         447.96      0.16 30.74   0.34   0.22  19.74
3      female       399.73      0.50 21.75   0.90   0.80  16.12
       male         455.52      0.14 26.51   0.50   0.22  12.66

Aqui agregamos os dados por Pclass e Sex e em todas as colunas numéricas foi calculada a média. Se quiséssemos a média de apenas uma coluna poderíamos adicioná-la ao final da nossa sentença:

In [37]:
# Queremos apenas a média de idade considerando a classe e o sexo
df.groupby(["Pclass", "Sex"]).mean()[['Age']]

Age
Pclass Sex         
1      female 34.61
       male   41.28
2      female 28.72
       male   30.74
3      female 21.75
       male   26.51

Ou de modo mais eficiente:

In [39]:
df.groupby(["Pclass", "Sex"])[['Age', 'Fare']].mean()

Age   Fare
Pclass Sex                
1      female 34.61 106.13
       male   41.28  67.23
2      female 28.72  21.97
       male   30.74  19.74
3      female 21.75  16.12
       male   26.51  12.66

In [61]:
df.groupby(["Pclass", "Sex"])[['Age', 'Fare']].mean()

Age   Fare
Pclass Sex                
1      female 34.61 106.13
       male   41.28  67.23
2      female 28.72  21.97
       male   30.74  19.74
3      female 21.75  16.12
       male   26.51  12.66

In [56]:
df.groupby(["Pclass", "Sex"])[['Age', 'Fare']].mean().index

MultiIndex([(1, 'female'),
            (1,   'male'),
            (2, 'female'),
            (2,   'male'),
            (3, 'female'),
            (3,   'male')],
           names=['Pclass', 'Sex'])

Note que `df.groupby('A').colname.mean()` é mais eficiente que `df.groupby('A').mean().colname` pois a agregação só será realizada na coluna de interesse (colname).

Quando queremos aplicar mais de uma operação chamamos o método `.agg()`

In [43]:
df.groupby(["Pclass"]).agg(['mean','max','min']).columns

MultiIndex([('PassengerId', 'mean'),
            ('PassengerId',  'max'),
            ('PassengerId',  'min'),
            (   'Survived', 'mean'),
            (   'Survived',  'max'),
            (   'Survived',  'min'),
            (        'Age', 'mean'),
            (        'Age',  'max'),
            (        'Age',  'min'),
            (      'SibSp', 'mean'),
            (      'SibSp',  'max'),
            (      'SibSp',  'min'),
            (      'Parch', 'mean'),
            (      'Parch',  'max'),
            (      'Parch',  'min'),
            (       'Fare', 'mean'),
            (       'Fare',  'max'),
            (       'Fare',  'min')],
           )

In [42]:
df.groupby(["Pclass"]).agg(['mean','max','min'])

PassengerId          Survived           Age            SibSp          \
              mean  max min     mean max min  mean   max  min  mean max min   
Pclass                                                                        
1           461.60  890   2     0.63   1   0 38.23 80.00 0.92  0.42   3   0   
2           445.96  887  10     0.47   1   0 29.88 70.00 0.67  0.40   3   0   
3           439.15  891   1     0.24   1   0 25.14 74.00 0.42  0.62   8   0   

       Parch          Fare              
        mean max min  mean    max  min  
Pclass                                  
1       0.36   4   0 84.15 512.33 0.00  
2       0.38   3   0 20.66  73.50 0.00  
3       0.39   6   0 13.68  69.55 0.00

Para operações distintas em colunas distintas passamos um dicionário com o nome da coluna como chave e a operação como valor

In [50]:
import numpy as np
df.groupby(['Pclass']).agg({'Embarked': pd.Series.mode, 'Fare': 'mean'})

Embarked  Fare
Pclass               
1             S 84.15
2             S 20.66
3             S 13.68

Reparem que a coluna utilizada no `groupby` virou um index do nosso df. Para convertê-la em coluna novamente temos duas formas: <br>
  1. chamar o parâmetro `as_index=False` dentro do `groupby`
  2. aplicar `.reset_index()` ao final da sentença

In [55]:
df.groupby(['Pclass']).agg({'Embarked': pd.Series.mode, 'Fare': 'mean'}).index

Int64Index([1, 2, 3], dtype='int64', name='Pclass')

In [57]:
type(df.groupby(['Pclass']).agg({'Embarked': pd.Series.mode, 'Fare': 'mean'}))

pandas.core.frame.DataFrame

In [58]:
# exemplo com as_index = False
df.groupby(['Pclass'], as_index=False).agg({'Embarked': pd.Series.mode, 'Fare': 'mean'})

Pclass Embarked  Fare
0       1        S 84.15
1       2        S 20.66
2       3        S 13.68

In [59]:
# exemplo com .reset_index()
df.groupby(['Pclass']).agg({'Embarked': pd.Series.mode, 'Fare': 'mean'}).reset_index()

Pclass Embarked  Fare
0       1        S 84.15
1       2        S 20.66
2       3        S 13.68

In [60]:
df.groupby(["Pclass", "Sex"])[['Age', 'Fare']].mean().reset_index()

Pclass     Sex   Age   Fare
0       1  female 34.61 106.13
1       1    male 41.28  67.23
2       2  female 28.72  21.97
3       2    male 30.74  19.74
4       3  female 21.75  16.12
5       3    male 26.51  12.66

In [62]:
df.groupby(["Pclass", "Sex"], as_index=False)[['Age', 'Fare']].mean()

Pclass     Sex   Age   Fare
0       1  female 34.61 106.13
1       1    male 41.28  67.23
2       2  female 28.72  21.97
3       2    male 30.74  19.74
4       3  female 21.75  16.12
5       3    male 26.51  12.66

_____________
_____________
**Exercício:** Existe diferença de sobrevivência por portão de embarque? E diferença no preço do ticket? Porque você acha que tem essa diferença?

In [71]:
# Resposta
df.groupby(["Embarked", 'Pclass'])[["Fare"]].mean()


Fare
Embarked Pclass       
C        1      104.72
         2       25.36
         3       11.21
Q        1       90.00
         2       12.35
         3       11.18
S        1       70.36
         2       20.33
         3       14.64

In [66]:
df.groupby(["Embarked", 'Pclass'])[["Survived", "Age"]].agg(['mean', 'count'])

Survived         Age      
                    mean count  mean count
Embarked Pclass                           
C        1          0.69    85 38.03    74
         2          0.53    17 22.77    15
         3          0.38    66 20.74    41
Q        1          0.50     2 38.50     2
         2          0.67     3 43.50     2
         3          0.38    72 25.94    24
S        1          0.58   127 38.15   108
         2          0.46   164 30.39   156
         3          0.19   353 25.70   290

In [69]:
df.corr()

PassengerId  Survived  Pclass   Age  SibSp  Parch  Fare
PassengerId         1.00     -0.01   -0.04  0.04  -0.06  -0.00  0.01
Survived           -0.01      1.00   -0.34 -0.08  -0.04   0.08  0.26
Pclass             -0.04     -0.34    1.00 -0.37   0.08   0.02 -0.55
Age                 0.04     -0.08   -0.37  1.00  -0.31  -0.19  0.10
SibSp              -0.06     -0.04    0.08 -0.31   1.00   0.41  0.16
Parch              -0.00      0.08    0.02 -0.19   0.41   1.00  0.22
Fare                0.01      0.26   -0.55  0.10   0.16   0.22  1.00

______________
_____________

E se quiséssemos criar uma coluna nova que contenham o valor médio do Fare por Pclass?

### Criando coluna com dado agregado

Queremos que todas as pessoas da primeira classe tenham o valor 84.15 nessa nova coluna, todas da segunda classe tenham o valor 20.66 e da terceira classe 13.67. <br>
Podemos tentar:

In [72]:
df.groupby('Pclass')[["Fare"]].mean()

Fare
Pclass      
1      84.15
2      20.66
3      13.68

In [73]:
df["Fare_Mean"] = df.groupby('Pclass')["Fare"].mean()

df.head(7)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male 54.00      0   

   Parch            Ticket  Fare Cabin Embarked  Fare_Mean  
0      0         A/5 21171  7.25   NaN        S        NaN  
1      0          PC 17599 71.28   C85        C      84.15  
2      0  STON/O2. 3101282  7.92   NaN        S      20.66  
3      0            113803 53.10  C123        S      13.68  
4      0            373450  8.05   NaN        S        NaN  
5      0            330877  8.46   NaN        Q        NaN  
6      0             17463 51.86   E46        S        NaN

Xiiii... deu ruim...
<br>
<br>


#### Transformação dos dados
Ao aplicarmos o método `.transform()` temos como retorno um objeto com o mesmo index do df de origem contendo a a transformação realizada para cada uma das linhas. Dessa forma podemos utilizar esse método e apenas criar uma coluna nova no nosso df.
<br>

Ele será muito **útil na criação de novas features** para os modelos.

In [75]:
df.groupby('Pclass')[["Fare"]].mean()

Fare
Pclass      
1      84.15
2      20.66
3      13.68

In [74]:
df.groupby('Pclass')[["Fare"]].transform('mean')

Fare
0   13.68
1   84.15
2   13.68
3   84.15
4   13.68
..    ...
886 20.66
887 84.15
888 13.68
889 84.15
890 13.68

[891 rows x 1 columns]

In [76]:
df["Fare_Mean"] = df.groupby('Pclass')["Fare"].transform('mean')
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male 54.00      0   
7                     Palsson, Master. Gosta Leonard    male  2.00      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female 27.00      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female 14.00      1   

   Parch            Ticket  Fare Cabin Embarked  Fare_Mean  
0      0         A/5 21171  7.25   NaN        S      13.68  
1      0          PC 17599 71.28   C85        C      84.15  
2      0  STON/O2. 3101282  7.92   NaN        S      13.68  
3      0            113803 53.10  C123        S      84.15  
4      0            373450  8.05   NaN        S      13.68  
5      0            330877  8.46   NaN        Q      13.68  
6      0             17463 51.86   E46        S      84.15  
7      1            349909 21.07   NaN        S      13.68  
8      2            347742 11.13   NaN        S      13.68  
9      0            237736 30.07   NaN        C      20.66

Podemos aplicar tanto as operações mencionadas na agregação quanto uma função `lambda`:

In [77]:
def mult_2(x):
    return x*2

mult_2(4)

8

In [78]:
funcao_lambda = lambda x: 2*x
funcao_lambda(4)

8

In [79]:
funcao_lambda

<function __main__.<lambda>(x)>

In [80]:
df['variacao_max_min'] = df.groupby('Pclass')[["Fare"]].transform(lambda x: x.max() - x.min())
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male 54.00      0   
7                     Palsson, Master. Gosta Leonard    male  2.00      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female 27.00      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female 14.00      1   

   Parch            Ticket  Fare Cabin Embarked  Fare_Mean  variacao_max_min  
0      0         A/5 21171  7.25   NaN        S      13.68             69.55  
1      0          PC 17599 71.28   C85        C      84.15            512.33  
2      0  STON/O2. 3101282  7.92   NaN        S      13.68             69.55  
3      0            113803 53.10  C123        S      84.15            512.33  
4      0            373450  8.05   NaN        S      13.68             69.55  
5      0            330877  8.46   NaN        Q      13.68             69.55  
6      0             17463 51.86   E46        S      84.15            512.33  
7      1            349909 21.07   NaN        S      13.68             69.55  
8      2            347742 11.13   NaN        S      13.68             69.55  
9      0            237736 30.07   NaN        C      20.66             73.50

Ou até mesmo passar funções construídas:

In [81]:
def funcao_max_menos_min(x):
    return x.max() - x.min()

In [82]:
df.groupby('Pclass')[["Fare"]].transform(funcao_max_menos_min)

Fare
0    69.55
1   512.33
2    69.55
3   512.33
4    69.55
..     ...
886  73.50
887 512.33
888  69.55
889 512.33
890  69.55

[891 rows x 1 columns]

Também podemos preencher os valores nulos com a média de cada grupo

In [83]:
# verificando quantidade de nulos por coluna
df.isna().sum()

PassengerId           0
Survived              0
Pclass                0
Name                  0
Sex                   0
Age                 177
SibSp                 0
Parch                 0
Ticket                0
Fare                  0
Cabin               687
Embarked              2
Fare_Mean             0
variacao_max_min      0
dtype: int64

Para preencher os nulos utilizaremos o método `.fillna()` que vimos em aula:

In [86]:
df[['Age']].mean()

Age   29.70
dtype: float64

In [92]:
df.Age.fillna(df.Age.mean())

0     22.00
1     38.00
2     26.00
3     35.00
4     35.00
       ... 
886   27.00
887   19.00
888   29.70
889   26.00
890   32.00
Name: Age, Length: 891, dtype: float64

In [87]:
df['Age_sem_nulo'] = df.groupby(['Pclass'])[['Age']].transform(lambda x: x.fillna(x.mean()))

In [88]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male 22.00      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female 38.00      1   
2                             Heikkinen, Miss. Laina  female 26.00      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female 35.00      1   
4                           Allen, Mr. William Henry    male 35.00      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male 54.00      0   
7                     Palsson, Master. Gosta Leonard    male  2.00      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female 27.00      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female 14.00      1   

   Parch            Ticket  Fare Cabin Embarked  Fare_Mean  variacao_max_min  \
0      0         A/5 21171  7.25   NaN        S      13.68             69.55   
1      0          PC 17599 71.28   C85        C      84.15            512.33   
2      0  STON/O2. 3101282  7.92   NaN        S      13.68             69.55   
3      0            113803 53.10  C123        S      84.15            512.33   
4      0            373450  8.05   NaN        S      13.68             69.55   
5      0            330877  8.46   NaN        Q      13.68             69.55   
6      0             17463 51.86   E46        S      84.15            512.33   
7      1            349909 21.07   NaN        S      13.68             69.55   
8      2            347742 11.13   NaN        S      13.68             69.55   
9      0            237736 30.07   NaN        C      20.66             73.50   

   Age_sem_nulo  
0         22.00  
1         38.00  
2         26.00  
3         35.00  
4         35.00  
5         25.14  
6         54.00  
7          2.00  
8         27.00  
9         14.00

In [89]:
df.isna().sum()

PassengerId           0
Survived              0
Pclass                0
Name                  0
Sex                   0
Age                 177
SibSp                 0
Parch                 0
Ticket                0
Fare                  0
Cabin               687
Embarked              2
Fare_Mean             0
variacao_max_min      0
Age_sem_nulo          0
dtype: int64

In [84]:
# Conferindo o preenchimento de nulos
# idade média por Pclass
df.groupby(['Pclass'])[['Age']].mean()

Age
Pclass      
1      38.23
2      29.88
3      25.14

In [90]:
# selecionando a parte do df que tem idade nula
df[df.Age.isna()].head(10)

PassengerId  Survived  Pclass  \
5             6         0       3   
17           18         1       2   
19           20         1       3   
26           27         0       3   
28           29         1       3   
29           30         0       3   
31           32         1       1   
32           33         1       3   
36           37         1       3   
42           43         0       3   

                                              Name     Sex  Age  SibSp  Parch  \
5                                 Moran, Mr. James    male  NaN      0      0   
17                    Williams, Mr. Charles Eugene    male  NaN      0      0   
19                         Masselmani, Mrs. Fatima  female  NaN      0      0   
26                         Emir, Mr. Farred Chehab    male  NaN      0      0   
28                   O'Dwyer, Miss. Ellen "Nellie"  female  NaN      0      0   
29                             Todoroff, Mr. Lalio    male  NaN      0      0   
31  Spencer, Mrs. William Augustus (Marie Eugenie)  female  NaN      1      0   
32                        Glynn, Miss. Mary Agatha  female  NaN      0      0   
36                                Mamee, Mr. Hanna    male  NaN      0      0   
42                             Kraeff, Mr. Theodor    male  NaN      0      0   

      Ticket   Fare Cabin Embarked  Fare_Mean  variacao_max_min  Age_sem_nulo  
5     330877   8.46   NaN        Q      13.68             69.55         25.14  
17    244373  13.00   NaN        S      20.66             73.50         29.88  
19      2649   7.22   NaN        C      13.68             69.55         25.14  
26      2631   7.22   NaN        C      13.68             69.55         25.14  
28    330959   7.88   NaN        Q      13.68             69.55         25.14  
29    349216   7.90   NaN        S      13.68             69.55         25.14  
31  PC 17569 146.52   B78        C      84.15            512.33         38.23  
32    335677   7.75   NaN        Q      13.68             69.55         25.14  
36      2677   7.23   NaN        C      13.68             69.55         25.14  
42    349253   7.90   NaN        C      13.68             69.55         25.14

_________________________
_________________________
**Exercício:** Crie uma coluna com a média de Fare e outra com a média de idade para cada classe da coluna Survived. Você consegue fazer isso de uma única vez?

In [ ]:
# Resposta


_________________________
_________________________

## Cruzamento e concatenação de bases

Também é possível fazer **cruzamento de bases** com o pandas. 

Pra quem conhece SQL: esses são os joins!

Pra quem conhece Excel: essa é uma forma de fazer o procv!

Vamos supor que temos as notas de duas provas dos alunos separas em sheets diferentes do excel e queremos juntar essa notas em um único df.

In [98]:
# ler os dados de diferentes sheets do mesmo excel "notas.xlsx"
df1 = pd.read_excel("notas.xlsx", sheet_name="notas1")
df2 = pd.read_excel("notas.xlsx", sheet_name="notas2")

In [99]:
df1

RA    aluno  prova1
0   1     joão      10
1   4     leia      10
2   2    maria       9
3   3      han       8
4   5     luke       7
5   7  obi wan      10

In [100]:
df2

RA   aluno  prova2
0   1    joão      10
1   4    leia      10
2   2   maria       9
3   3     han       8
4   5    luke       7
5   6  anakin      10

Repare que temos alunos distintos nos dois df

Diferentes tipos de join

<img src="join_exemplo2.png" />
Fonte: https://towardsdatascience.com/python-pandas-dataframe-join-merge-and-concatenate-84985c29ef78

O pandas possui dois métodos específicos para trabalharmos com o join de colunas entre df: `.merge()` e `.join()`. O `.merge()` fornece mais flexibilidade de trabalho e iremos utilizar e ele.

### pd.merge()
pd.merge(
    left,
    right,
    how="inner",
    on=None,
    left_on=None,
    right_on=None,
    left_index=False,
    right_index=False,
    sort=True,
    suffixes=("_x", "_y")
)

In [103]:
df1.merge(df2[['RA','prova2']], how="outer", on="RA")

RA    aluno  prova1  prova2
0   1     joão   10.00   10.00
1   4     leia   10.00   10.00
2   2    maria    9.00    9.00
3   3      han    8.00    8.00
4   5     luke    7.00    7.00
5   7  obi wan   10.00     NaN
6   6      NaN     NaN   10.00

In [110]:
df1.merge(df2, how="outer", on="RA", suffixes=['_prova1','_prova2' ])

RA aluno_prova1  prova1 aluno_prova2  prova2
0   1         joão   10.00         joão   10.00
1   4         leia   10.00         leia   10.00
2   2        maria    9.00        maria    9.00
3   3          han    8.00          han    8.00
4   5         luke    7.00         luke    7.00
5   7      obi wan   10.00          NaN     NaN
6   6          NaN     NaN       anakin   10.00

In [105]:
df1.merge(df2, how="outer", on=["RA", "aluno"])

RA    aluno  prova1  prova2
0   1     joão   10.00   10.00
1   4     leia   10.00   10.00
2   2    maria    9.00    9.00
3   3      han    8.00    8.00
4   5     luke    7.00    7.00
5   7  obi wan   10.00     NaN
6   6   anakin     NaN   10.00

In [106]:
df1.merge(df2, how="left", on=["RA", "aluno"])

RA    aluno  prova1  prova2
0   1     joão      10   10.00
1   4     leia      10   10.00
2   2    maria       9    9.00
3   3      han       8    8.00
4   5     luke       7    7.00
5   7  obi wan      10     NaN

In [107]:
df1.merge(df2, how="right", on=["RA", "aluno"])

RA   aluno  prova1  prova2
0   1    joão   10.00      10
1   4    leia   10.00      10
2   2   maria    9.00       9
3   3     han    8.00       8
4   5    luke    7.00       7
5   6  anakin     NaN      10

In [108]:
df1.merge(df2, how="inner", on=["RA", "aluno"])

RA  aluno  prova1  prova2
0   1   joão      10      10
1   4   leia      10      10
2   2  maria       9       9
3   3    han       8       8
4   5   luke       7       7

### pd.concat()
Diferente do `.merge()` e `.join()` que operam apenas com colunas, com o `.concat()` podemos especificar se queremos **concatenar em linhas ou colunas**.
Na concatenação de colunas o `.concat()` somente considera o index dos df e, por isso, não podemos especificar colunas como feito com o `.merge()`.

`pd.concat(
    objs,
    axis=0,
    join="outer",
    ignore_index=False,
    keys=None,
    levels=None,
    names=None,
    verify_integrity=False,
    copy=True,
)`


In [ ]:
pd.concat([df1, df2], axis=1, join="inner")

Repare que ao concatenar diretamente pelo index ele juntou o aluno obi wan com o anakin. 

Ao concatenar dois df nas linhas, o `.concat()` irá considerar o nome das colunas. Se temos colunas com nomes distintos e utilizamos o parâmetro join='inner', ele irá ignorar essas colunas: 

In [111]:
pd.concat([df1, df2], axis=0, join="inner")

RA    aluno
0   1     joão
1   4     leia
2   2    maria
3   3      han
4   5     luke
5   7  obi wan
0   1     joão
1   4     leia
2   2    maria
3   3      han
4   5     luke
5   6   anakin

Para que ele considere todas as colunas utilizamos o argumento 
```python 
join="outer" 
```

In [112]:
pd.concat([df1, df2], join="outer")

RA    aluno  prova1  prova2
0   1     joão   10.00     NaN
1   4     leia   10.00     NaN
2   2    maria    9.00     NaN
3   3      han    8.00     NaN
4   5     luke    7.00     NaN
5   7  obi wan   10.00     NaN
0   1     joão     NaN   10.00
1   4     leia     NaN   10.00
2   2    maria     NaN    9.00
3   3      han     NaN    8.00
4   5     luke     NaN    7.00
5   6   anakin     NaN   10.00

## Exercícios

1. Considere a existência de três tabelas distintas:
* customer.csv : Possui a informação dos clientes em duas colunas: customer id  customer name
* products.csv : Conté informação dos produtos vendidos pela empresa em três colunas - p_id (product id), product (name) e price
* sales.csv : Contém informações das vendas realizadas em seis colunas - sale_id, c_id (customer id), p_id (product_id), qty (quantity sold), store (name)

Conhecendo as bases e utilizando os métodos de concatenação de bases responda:


a) Quais produtos não foram vendidos?

In [ ]:
import pandas as pd 
sales = pd.read_csv("data/sales.csv") 
products = pd.read_csv("data/products.csv") 

# Faz um merge priorizando as informações da tabela de produtos
my_data = pd.merge(sales, products, on=['p_id','product'],how='right')
 
# Seleciona todos as linhas em que o sale_id é nan
my_data = my_data[my_data['sale_id'].isna()]

# Seleciona os produtos que não foram vendidos
print(my_data.loc[:,'product'])

b) Quantos clientes não realizaram uma compra? 

In [ ]:
customers = pd.read_csv("data/customer.csv") 

# Faz um merge priorizando as informações da tabela de customers
my_data = pd.merge(sales, customers, on='c_id', how='right')

# Seleciona todos as linhas em que o sale_id é nan
my_data = my_data[my_data['sale_id'].isna()] 

# Seleciona os clientes que não realizaram compras
print(my_data.loc[:,'Customer']) 

c) Liste a quantidade vendida e o faturamento de cada produto 

In [ ]:
my_sum=pd.merge(sales, products, how='left', on=['p_id','product'])

# Calcula valor total de cada sale_id
my_sum['total_sale'] = my_sum['qty']*my_sum['price']

# Agrupa por produto e soma a quantidade vendida e o valor total da compra
my_sum.groupby(['product'])[['qty', 'total_sale']].sum()

d) Liste a quantidade vendida de cada produto por loja

In [ ]:
my_sale = sales.groupby(['product','p_id', 'store'])[['qty']].sum().reset_index()
my_sale

e) Qual loja teve maior faturamento?

In [ ]:
my_sum.groupby(['store'])[['total_sale']].sum().sort_values('total_sale').tail(1)

f) Qual produto foi o mais vendido?

In [ ]:
my_sum.groupby(['product'])[['qty']].sum().sort_values('qty').tail(1)

## Referências
https://pandas.pydata.org/docs/user_guide/groupby.html <br>
https://pandas.pydata.org/docs/user_guide/merging.html <br> 
https://towardsdatascience.com/python-pandas-dataframe-join-merge-and-concatenate-84985c29ef78 <br>
[When to use pandas transform function](https://towardsdatascience.com/when-to-use-pandas-transform-function-df8861aa0dcf) <br>
[Compara a performance entre várias formas de iterar em um df. Vai desde o for até apply e transform](https://youtu.be/rsyvErL0Bo8) <br>

## Material extra

### Outros parâmetros do groupby por default
* as_index
* sort
* dropna # exclui nans nas keys

<br> Em todas o default do python é True <br>
df.groupby('Pclass', sort=False)["Fare"].mean()

In [ ]:
# dropna
df_list = [[1, 2, 3], [1, None, 4], [2, 1, 3], [1, 2, 2]]
df_dropna = pd.DataFrame(df_list, columns=["a", "b", "c"])
df_dropna

In [ ]:
# Default ``dropna`` is set to True, which will exclude NaNs in keys
df_dropna.groupby(by=["b"], dropna=True).sum()

In [ ]:
df_dropna.groupby(by=["b"], dropna=False).sum()

Repare que podemos chamar qualquer função do `pd.Series` ou  do `numpy`

In [ ]:
df.groupby(["Survived"]).mean()

### Função Lambda
Uma função lambda nada mais é que uma **forma alternativa de declarar uma função**, de um jeito mais direto

In [ ]:
# função que retorna o dobro de um número usando def
def dobro(x):
    
    return 2*x

dobro(2)

In [ ]:
# função que retorna o dobro de um número usando lambda x
faz_dobro = lambda x: 2*x

In [ ]:
faz_dobro(6)

### Apply
O método `.apply()` recebe uma função como input e aplica ela para todo o df como se fosse um loop. Se você quiser que essa função seja aplicada ao longo das colunas deve considerar axis=0 e ao longo das linhas axis=1)

In [ ]:
df.groupby(['Pclass']).apply(lambda x: x.describe())

Uma grande funcionalidade do pandas é que com o método `apply()` podemos aplicar uma **função** (muitas vezes, uma **função lambda**) a uma coluna ou linha de um DataFrame



Vamos selecionar a coluna de idades...

In [ ]:
df["Age"]

Aplicando uma função lambda **a todos os elementos da coluna**, ou seja, **à todas as linhas da tabela, daquela coluna específica**:

Tomando cada idade + 2, usando a função lambda definida.

Essa função lambda é equivalente a:

```python

def funcao(x):

    return x + 2
```

In [ ]:
df["Age"].apply(lambda x: x + 2)

In [ ]:
def funcao(x):
    return x + 2

df.Age.apply(funcao)

In [ ]:
df.Age.transform(funcao)

Um outro exemplo:

In [ ]:
# função: transforma todos os números em string, e concatena "!!!!!!!!!" à string
df["Age"].apply(lambda x: str(x) + "!!!!!!!!!")

Vamos usar uma função lambda para **extrair o sobrenome** dos nomes dos passageiros

Pra extrarir o sobrenome, note que este está separada do resto do nome por vírgula.

Para perceber isso, dê uma olhada na coluna de nomes:

In [ ]:
df["Name"]

Portanto, podemos usar a função para strings `split(",")`, com quebra na vírgula, e depois selecionar o primeiro elemento da lista gerada!

Vamos aproveitar e **criar uma nova coluna da base**, com os sobrenomes!

In [ ]:
df["Surname"] = df["Name"].apply(lambda x: x.split(",")[0])

In [ ]:
df["Surname"]

### Apply com funções

E se quisessemos comparar o quanto cada passageiro pagou a mais ou a menos da média do Fare?

In [ ]:
def f(group):
    return pd.DataFrame({'Fare_original': group,
                         'Fare_variacao': group - group.mean()})

df[['Fare_original','Fare_variacao']] = df.groupby('Pclass')['Fare'].apply(f)

In [ ]:
df.head()

#### Transform X Apply
Com uma função de agregação o `.transform()` retorna um df que tem a mesma quantidade de linhas que o df original enquanto o `.apply` retorna o agregado por grupos.

### Filtros
O filtro retorna apenas um subset do nosso df. Aqui podemos aplicar filtros mais elaborados do que os vistos na última aula. <br>
Podemos, por exemplo, eliminar categorias do df que possuem apenas alguns elementos:

In [ ]:
df.SibSp.value_counts()

In [ ]:
df.shape

In [ ]:
def filter_func(x):
    return x['Fare'] - x.Fare_Mean < 100

# df_filter = df.groupby(['SibSp']).filter(lambda x: filter_func(x))

df_filter = df.groupby(['SibSp']).filter(lambda x: len(x) >20)
df_filter.shape

In [ ]:
df_filter.SibSp.value_counts()

Vamos supor que antes de afundar o titanic, o time de hapiness quisesse promover uma jogatina para os grupos (segmentado por classe e sexo) que possuem idade média acima de 30 anos.

In [ ]:
df.groupby(['Pclass','Sex'])[['Age']].mean()

como podemos filtrar nosso df para termos apenas os passageiros que pertecem a essas segmentações escolhidas?

In [ ]:
df.groupby(['Pclass','Sex']).filter(lambda x: x['Age'].mean()>30)